### PR TITLE
Sha256f Opcode

### PR DESCRIPTION
**Pull request by hkalodner**
- SHA256F added to cpp avm
- SHA256F added to ethbrige
- Add SHA256F benchmarking and one step proof test
- SHA256F gas cost is 250

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

File: docs/AVM_Specification.md [hunk 1]
```diff
@@ -254,3 +255,13 @@ The function carries out three steps.
 1. Using the input array of seven Integers, derive a 200-byte string. For i = 0 through 5 inclusive, bytes _32i_ through _32i+31_ of the result are set equal to the 32-byte big-endian representation of _input[i]_. Bytes 192-199 of the result are set equal to the 8-byte big-endian representation _input[6] % (2\*\*64)_.
 2. Using the result of the Step 1 as input, compute the function KECCAK-p[1600, 24] as defined in the SHA-3 standard (FIPS PUB 202, dated August 2015), producing a 200-byte string as output.
 3. Using the result of Step 2 is input, derive an output array of seven Integers. This array is the result produced by the `keccakf`instruction. For i = 0 through 5 inclusive, _output[i]_ is set equal to bytes _32i_ through _32i+31_ of the result of Step 2 (interpreted as a 32-byte big-endian Integer). output[6] is set equal to bytes 192-199 of the result of Step 2 (interpreted as an 8-byte big-endian Integer in the range 0 through _(2\*\*64)-1)_).
+
+### Definition of `sha256f`
+
+The `sha256f` function is computer by the `sha256f` instruction. The function takes 3 Integers as input, and produces, and produces an Integer as output.
```
Comment (edfelten) [inline docs/AVM_Specification.md:261]: type: "computer" should be "computed"

File: docs/AVM_Specification.md [hunk 2]
```diff
@@ -254,3 +255,13 @@ The function carries out three steps.
...
+
+The function carries out three steps.
+
+1. The 32 byte little-endian represention of the first item from the stack is interpretted as the digest and the 32 byte representations of the second and third items from the stack are concatenated and interpretted as the input array.
```
Comment (edfelten) [inline docs/AVM_Specification.md:265]: typo: "interpreted" should be "interpreted". (appears twice in this paragraph)

File: packages/arb-bridge-eth/contracts/libraries/Precompiles.sol [hunk 3]
```diff
@@ -176,4 +176,149 @@ library Keccak {
...
+        uint32 f,
+        uint32 g
+    ) internal pure returns (uint32) {
+        return ((e & f) ^ ((~e) & g));
+    }
+
+    // SHA256 compression function that operates on a 512 bit chunk
+    // Note that the input must be padded by the caller
+    // For the initial chunk, the initial values from the SHA256 spec should be passed in as hashState
+    // For subseuqent rounds, hashState is the output from the previous round
```
Comment (edfelten) [inline packages/arb-bridge-eth/contracts/libraries/Precompiles.sol:195]: typo: "subseuqent" should be "subsequent"

**Review by edfelten — CHANGES_REQUESTED** [review]
Looks good, except for a few typos.

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  3
         

Clones added
============
- packages/arb-validator-core/ethbridgetestcontracts/precompilestester.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6165536&bid=20135898)

**Review by edfelten — APPROVED** [review]

**Pull request merged**